Add PWR037: Potential precision loss in call to mathematical function

alvrogd opened the pull request:
Add entry and benchmark for a check that was already defined. My results:

```txt
$ lscpu
  Model name:             13th Gen Intel(R) Core(TM) i7-13700H

$ gcc --version
gcc (Debian 12.2.0-14) 12.2.0

$ ./run-benchmarks.py --check PWR037
-----------------------------------------------------------
Benchmark                 Time             CPU   Iterations
-----------------------------------------------------------
PWR037 C Example      15610 us        15607 us           45
PWR037 C Fixed        22523 us        22517 us           31
```

Note that the performance impact is expected when performing higher-precision calculations. Still, the check helps ensure the accuracy and reliability of the code.

inaki-amatria commented on Checks/PWR037/benchmark/solution.c:
```diff
@@ -0,0 +1,18 @@
+// PWR037: Potential precision loss in call to mathematical function
+
+#include <math.h>
```
Ensure that `add_benchmark` links with `libm`. It seems this linkage might be missing!

inaki-amatria replied:
You may cherry-pick this commit https://github.com/codee-com/open-catalog/pull/58/commits/ec8a029315b48ac60d5f70abcf88e75f74bea222

alvrogd replied:
Thanks for the catch and the reference! I will cherry-pick that :)

igarcia0 commented on Checks/PWR037/README.md:
```diff
@@ -1,3 +1,115 @@
+
+Note the error in the result beyond the seventh decimal place. The accurate
+approximation for the square root of 2 would be `1.41421 35623 73095`. The
+mismatch corresponds to the 7-digit precision limit of the `float` data type on
```
Nit: FP32 may give from 6 to 9 significant decimal digits.

alvrogd replied:
If I'm not mistaken, the 6 to 9 discussion is related to what's needed for reliable back-and-forth conversion between decimal representations and binary representations. And beyond that, the generally accepted maximum accuracy in decimal places when doing floating point arithmetic would still be 7. See https://en.wikipedia.org/wiki/Single-precision_floating-point_format#IEEE_754_standard:_binary32 for more details.

In any case, if you don't mind, I believe this would be too technical for the purpose of the check entry, so I believe it would be best to leave it out :)

igarcia0 reviewed (approved):
LGTM

alvrogd commented on Checks/PWR037/README.md:
```diff
@@ -1,3 +1,115 @@
+
+Using a function intended for a smaller data type, such as `sqrtf()` with a
+`double` input, causes an implicit narrowing of the input value. This implicit
+type conversion is allowed by the compiler, but results in a loss of precision
```
Indeed! I've updated the text.

alvrogd commented on Checks/PWR037/README.md:
````diff
@@ -1,3 +1,115 @@
+Square root of 2 is: 1.41421 35623 73095
+```
+
+> [!IMPORTANT]
+> While higher-precision calculations may come with a slight performance cost,
+> they help ensure the accuracy and reliability of the code. Instead, if the
+> intended choice is to use the lower-precision calculation, adjust the data
+> type of the variable to ensure it matches the called function.
````
That's a good way to frame it. I've updated this note. Let me know what you think!

alvrogd commented:
> Looks great!

Since your review was published at the same time that I merged this branch, I've addressed your suggestions in a new follow-up PR #59!
